### PR TITLE
Implementation of iProov federated authenticator

### PR DESCRIPTION
**Pull request by lashinijay**
## Automation tests
 - Unit tests 
   > Code coverage information
 - Integration tests
   > Details about the test cases and coverage

## Security checks
 - Followed secure coding standards in http://wso2.com/technical-reports/wso2-secure-engineering-guidelines? yes/no
 - Ran FindSecurityBugs plugin and verified report? yes/no
 - Confirmed that this PR doesn't commit any keys, passwords, tokens, usernames, or other secrets? yes/no

### CODE REVIEW

**Review by ThaminduR — COMMENTED** [review]
Add copyright headers

File: components/org.wso2.carbon.identity.application.authenticator.iproov.common/src/main/java/org/wso2/carbon/identity/application/authenticator/iproov/common/constants/IproovAuthenticatorConstants.java [hunk 1]
```diff
@@ -0,0 +1,238 @@
...
+
+    /**
+    * Error messages for iProov authenticator.
+    */
+    public enum ErrorMessages {
+
+        AUTHENTICATION_FAILED_REDIRECTING_LOGIN_FAILURE("65001",
+                "Authentication failed when redirecting the user to the login page."),
+        USER_NOT_FOUND("65002", "User not found in the system. Please contact your system administrator."),
+        USER_ACCOUNT_LOCKED("65003", "User account i Please contact your system administrators locked."),
```
Comment (ThaminduR) [inline components/org.wso2.carbon.identity.application.authenticator.iproov.common/src/main/java/org/wso2/carbon/identity/application/authenticator/iproov/common/constants/IproovAuthenticatorConstants.java]: improve error message

File: components/org.wso2.carbon.identity.application.authenticator.iproov.common/src/main/java/org/wso2/carbon/identity/application/authenticator/iproov/common/exception/IproovAuthenticatorClientException.java [hunk 2]
```diff
@@ -0,0 +1,17 @@
+package org.wso2.carbon.identity.application.authenticator.iproov.common.exception;
+
+/**
+ * Iproov Authenticator Client Exception.
+ */
+public class IproovAuthenticatorClientException extends IproovAuthenticatorException {
+
+    private final String code;
+
+    private String description;
```
Comment (ThaminduR) [inline components/org.wso2.carbon.identity.application.authenticator.iproov.common/src/main/java/org/wso2/carbon/identity/application/authenticator/iproov/common/exception/IproovAuthenticatorClientException.java]:
```suggestion
    private final String description;
```

File: components/org.wso2.carbon.identity.application.authenticator.iproov.common/src/main/java/org/wso2/carbon/identity/application/authenticator/iproov/common/exception/IproovAuthenticatorClientException.java [hunk 3]
```diff
@@ -0,0 +1,17 @@
...
+    public IproovAuthenticatorClientException(String message, String description, String code, Throwable cause) {
```
Comment (ThaminduR) [inline components/org.wso2.carbon.identity.application.authenticator.iproov.common/src/main/java/org/wso2/carbon/identity/application/authenticator/iproov/common/exception/IproovAuthenticatorClientException.java:29]:
```suggestion

    public IproovAuthenticatorClientException(String message, String description, String code, Throwable cause) {
```

File: components/org.wso2.carbon.identity.application.authenticator.iproov.common/src/main/java/org/wso2/carbon/identity/application/authenticator/iproov/common/exception/IproovAuthenticatorException.java [hunk 4]
```diff
@@ -0,0 +1,18 @@
+package org.wso2.carbon.identity.application.authenticator.iproov.common.exception;
+
+import org.wso2.carbon.identity.base.IdentityException;
+
+/**
+ * Iproov Authenticator Exception.
+ */
+public class IproovAuthenticatorException extends IdentityException {
+    public IproovAuthenticatorException(String message) {
```
Comment (ThaminduR) [inline components/org.wso2.carbon.identity.application.authenticator.iproov.common/src/main/java/org/wso2/carbon/identity/application/authenticator/iproov/common/exception/IproovAuthenticatorException.java:28]:
```suggestion

    public IproovAuthenticatorException(String message) {
```

File: components/org.wso2.carbon.identity.application.authenticator.iproov.common/src/main/java/org/wso2/carbon/identity/application/authenticator/iproov/common/web/IproovAuthorizationAPIClient.java [hunk 5]
```diff
@@ -0,0 +1,201 @@
...
+     * @param tokenPath The token path of the iProov server.
+     * @param apiKey The API key of the created iProov service provider.
+     * @param secret The secret of the created iProov service provider.
+     * @param userId The user ID.
+     * @param token The token from the iProov server.
+     * @return The status of the verification token.
+     * @throws IproovAuthnFailedException If an error occurred when validating the verification token from the iProov
+     * server.
+     */
+    public static String validateVerification(String baseUrl, String tokenPath, String apiKey, String secret,
```
Comment (ThaminduR) [inline components/org.wso2.carbon.identity.application.authenticator.iproov.common/src/main/java/org/wso2/carbon/identity/application/authenticator/iproov/common/web/IproovAuthorizationAPIClient.java]: let's return a boolean

File: components/org.wso2.carbon.identity.application.authenticator.iproov.common/src/main/java/org/wso2/carbon/identity/application/authenticator/iproov/common/web/IproovAuthorizationAPIClient.java [hunk 6]
```diff
@@ -0,0 +1,201 @@
...
+     * @param userId The user ID.
+     * @throws IproovAuthnFailedException If an error occurred when removing the iProov user profile from the iProov
+     * server.
+     */
+    public static void removeIproovUserProfile(String baseUrl, String apiKey, String clientId, String secret,
+                                               String userId) throws IproovAuthnFailedException {
+
+        try {
+            URIBuilder uriBuilder = new URIBuilder(baseUrl);
+            uriBuilder.setPath(IproovAuthenticatorConstants.TokenEndpoints.IPROOV_DELETE_USER_PATH + "/" + userId);
```
Comment (ThaminduR) [inline components/org.wso2.carbon.identity.application.authenticator.iproov.common/src/main/java/org/wso2/carbon/identity/application/authenticator/iproov/common/web/IproovAuthorizationAPIClient.java]: Can't we include "/" in the constant as well

File: components/org.wso2.carbon.identity.application.authenticator.iproov.common/src/main/java/org/wso2/carbon/identity/application/authenticator/iproov/common/web/IproovWebUtils.java [hunk 7]
```diff
@@ -0,0 +1,153 @@
...
+import java.util.ArrayList;
+import java.util.List;
+
+/**
+ * Iproov Web Utils.
+ */
+public class IproovWebUtils {
+
+    private static final Log LOG = LogFactory.getLog(IproovWebUtils.class);
+    private IproovWebUtils() {
```
Comment (ThaminduR) [inline components/org.wso2.carbon.identity.application.authenticator.iproov.common/src/main/java/org/wso2/carbon/identity/application/authenticator/iproov/common/web/IproovWebUtils.java:60]:
```suggestion

    private IproovWebUtils() {
```

File: components/org.wso2.carbon.identity.application.authenticator.iproov.common/src/main/java/org/wso2/carbon/identity/application/authenticator/iproov/common/web/IproovWebUtils.java [hunk 8]
```diff
@@ -0,0 +1,153 @@
...
+            IproovAuthenticatorServerException, IproovAuthnFailedException {
+
+        try (CloseableHttpClient httpClient = HttpClientBuilder.create().useSystemProperties().build()) {
+
+            String tokenEndpoint = baseUrl + "/" + apiKey +
+                    IproovAuthenticatorConstants.TokenEndpoints.ACCESS_TOKEN_PATH;
+            HttpPost httpPost = new HttpPost(tokenEndpoint);
+            buildBasicAuthHeader(httpPost, clientId, clientSecret);
+            List<BasicNameValuePair> urlParameters = new ArrayList<>();
+            urlParameters.add(new BasicNameValuePair("grant_type",
```
Comment (ThaminduR) [inline components/org.wso2.carbon.identity.application.authenticator.iproov.common/src/main/java/org/wso2/carbon/identity/application/authenticator/iproov/common/web/IproovWebUtils.java]: use constants

File: components/org.wso2.carbon.identity.application.authenticator.iproov.common/src/main/java/org/wso2/carbon/identity/application/authenticator/iproov/common/web/IproovWebUtils.java [hunk 9]
```diff
@@ -0,0 +1,153 @@
...
+                if (response.getStatusLine().getStatusCode() == HttpStatus.SC_OK) {
+                    InputStreamReader reader = new InputStreamReader(response.getEntity().getContent(),
+                            StandardCharsets.UTF_8);
+                    BufferedReader in = new BufferedReader(reader);
+                    String json = in.readLine();
+                    // Parse the Json response and retrieve the Access Token.
+                    Gson gson = new Gson();
+                    JsonObject tokenDetails = gson.fromJson(json, JsonObject.class);
+                    in.close();
+                    return tokenDetails.get("access_token").getAsString();
```
Comment (ThaminduR) [inline components/org.wso2.carbon.identity.application.authenticator.iproov.common/src/main/java/org/wso2/carbon/identity/application/authenticator/iproov/common/web/IproovWebUtils.java]: use constants. check other places as well.

File: components/org.wso2.carbon.identity.application.authenticator.iproov.common/src/main/java/org/wso2/carbon/identity/application/authenticator/iproov/common/web/IproovWebUtils.java [hunk 10]
```diff
@@ -0,0 +1,153 @@
...
+                }
+                throw new IproovAuthenticatorServerException("Error occurred while retrieving the access token. " +
```
Comment (ThaminduR) [inline components/org.wso2.carbon.identity.application.authenticator.iproov.common/src/main/java/org/wso2/carbon/identity/application/authenticator/iproov/common/web/IproovWebUtils.java:160]: better to add a debug log here with the error received from iProov

File: components/org.wso2.carbon.identity.application.authenticator.iproov.rest/org.wso2.carbon.identity.application.authenticator.iproov.rest.dispatcher/src/main/resources/ErrorMappings.properties
Comment (ThaminduR) [inline components/org.wso2.carbon.identity.application.authenticator.iproov.rest/org.wso2.carbon.identity.application.authenticator.iproov.rest.dispatcher/src/main/resources/ErrorMappings.properties:1]: do we need the empty file?

File: components/org.wso2.carbon.identity.application.authenticator.iproov.common/pom.xml [hunk 11]
```diff
@@ -0,0 +1,83 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (ashensw) [inline components/org.wso2.carbon.identity.application.authenticator.iproov.common/pom.xml:1]: Add licence.

File: components/org.wso2.carbon.identity.application.authenticator.iproov.common/src/main/java/org/wso2/carbon/identity/application/authenticator/iproov/common/constants/IproovAuthenticatorConstants.java [hunk 12]
```diff
@@ -0,0 +1,242 @@
+package org.wso2.carbon.identity.application.authenticator.iproov.common.constants;
```
Comment (ashensw) [inline components/org.wso2.carbon.identity.application.authenticator.iproov.common/src/main/java/org/wso2/carbon/identity/application/authenticator/iproov/common/constants/IproovAuthenticatorConstants.java:19]: Add licence.  Check all the files and Add licence.

**Review by ashensw — APPROVED** [review]

**Pull request merged**